# BIN420

Bioinformatics for functional metagenomics (BIN420) II

Dates: 4-8 November 2019

Location: NMBU Ås

Organiser: Phillip B. Pope

Credits: 5 ECTS. NMBU students can register for credits here. Students from other institutions must first register here (EE-PhD) before September 1, and then register for credits here. Read more here.

Evaluation: A project report will be compulsory and must be submitted two weeks after the course. The report will be marked as passed/not passed.

Registration is closed

Course website can be found here

Lecturers: Lars G. Snipen (NMBU), Magnus Ø. Arntzen (NMBU), Live H. Hagen (NMBU), J. Chris Gaby (NMBU), and Francesco Delogu (NMBU)

 

Invited lecturer:

Patrick May (University of Luxembourg)

Benoit Kunath (University of Luxembourg)

 

Course description:

Microbial communities are renowned for the influences they exert in nature as well as industrial applications. They encompass an extraordinary level of species complexity that is invaluable to the overall function of the community. However, our understanding is severely restricted due to the inherent species complexity and the fact that the majority of microbes that exist in nature cannot be cultivated.

This course will introduce, explore and assess the vast array of sequencing technology and bioinformatic methods that are available to address these core issues. The course will include an array of contrasting tools to decrypt microbial communities, including those that assess community structure (metagenomics, predictive genome-reconstruction) and function (metatranscriptomics, metaproteomics).

 

Course program:

The course combines lectures (before lunch) and computer labs after lunch each day. This is an intensive course over five days:

Session 1 – Shotgun sequencing and analysis (Day 1-3)

* Design of experiments, possibilities and limitations with respect to technologies and bioinformatics tools. Deciding which software influences initial project design (i.e how many samples, how much sequencing, etc). Combining technologies, coverage calculations, enriched cultures.

* Metagenome sequence assembly. What is different from standard genome assembly?

* Taxonomic binning/classification of assembled metagenomic data.

* Gene prediction and functional annotation.

Session 4 – Functional meta-omics (Day 4-5)

* Metatranscriptomic data analysis.

* Metaproteomic data analysis.

* How to interpret the function of uncultured microorganisms (metaproteomic and metatranscriptomics).

 

Prerequisites:

* Ability to design experiments and select appropriate methods and/or software

* Explain the microbial population structure within a microbial community

* Ability to perform and assess assembly and taxonomic binning methods as well as interpret output quality

* Ability to combine the output from different omics methods to interpret the predicted function of uncultured microorganisms within a microbial community

* Explain the shortcomings about these types of analyses

 

Learning outcomes and competence

Familiarity with Linux and a programming language (preferably Python and/or R) is expected. Basic knowledge in microbiology is also required.
